### PR TITLE
Refactor Normal and MvNormal

### PR DESCRIPTION
**Pull request by gBokiau**
This PR refactor Normal and MvNormal: 
- clean up unused import and outdated comments
- switching to using functions in theano 1.0 release
- use (optimized) logp_sum where it makes sense

### CODE REVIEW

**Comment by ferrine** [comment]
Thanks for the PR, this is a random fail on Travis, nevermind. Do you see any more code style nitpicks?

**Comment by gBokiau** [comment]
Just PR'ing as I come across them. If there's a preference for less commits I'm happy to keep this open for a while in the eventuality of more crossing my path. 

I'm toying with the idea of harmonising all the covariance/precision/cholesky parameters declaration into classes of their own to unclutter model declaration and remove the code duplication. I'll PR separately if that reaches anywhere, however there's a good chance I bump into other tidbits such as the above while doing so.

File: pymc3/distributions/multivariate.py [hunk 1]
```diff
@@ -50,30 +50,27 @@ def __init__(self, mu=None, cov=None, chol=None, tau=None, lower=True,
...
-            self.k = cov.shape[0]
             self._cov_type = 'cov'
             cov = tt.as_tensor_variable(cov)
-            if cov.ndim != 2:
-                raise ValueError('cov must be two dimensional.')
-            self.chol_cov = cholesky(cov)
+            try:
+                self.chol_cov = cholesky(cov)
+            except ValueError:
+                raise ValueError('cov must be two dimensional.') from None
```
Comment (junpenglao) [inline pymc3/distributions/multivariate.py]: This is causing error in py2.7

File: pymc3/distributions/multivariate.py [hunk 2]
```diff
@@ -88,13 +85,11 @@ def _quaddist(self, value):
...
 
-        if self._cov_type == 'cov':
-            # Use this when Theano#5908 is released.
-            # return MvNormalLogp()(self.cov, delta)
-            dist, logdet, ok = self._quaddist_cov(delta)
-        elif self._cov_type == 'tau':
+        if self._cov_type == 'tau':
             dist, logdet, ok = self._quaddist_tau(delta)
         else:
+            # Use this when Theano#5908 is released.
```
Comment (junpenglao) [inline pymc3/distributions/multivariate.py]: Pretty sure theano#5908 is released. Is there way to incorporate this comment? cc @aseyboldt

**Comment by gBokiau** [comment]
Ah, I just realize I have gotten it wrong, I thought MvNormalLogp was somewhere in Theano. Will have another go.
I also think there's a couple places we could leverage Theano Views and inplace operations that are disabled by default in their convenience classes (e.g. in `tt.diagonal`). I have a suspicion a lot of solves could be done inplace.

**Comment by junpenglao** [comment]
Thanks for the work - I edit the PR title and comment.

**Comment by gBokiau** [comment]
One question about the Cholesky/logp implementation. If my understanding is correct, the decomposition of any positive definite (PD) matrix would have positive diagonals. Theano/Scipy returns a matrix of NaNs if the covariance matrix is not PD. This *is* different from the current implementation -- ie it returns an identity with only [0,0] set to NaN -- but I think the difference is trivial. 

Hence checking the diagonal has no purpose and logp should return a matrix of -Inf if the Cholesky fails. Is that correct?

--- 
Edit : Having now spent a few hours with that bit of code, it does look like it was (somewhat) misconceived in the sense that it applied logic pertaining to `chol` parameters whereas it was designed to handle `cov`. I've tried correcting that and implementing both parameters. I'm now out to test if it works as expected (commits below if others want to have a go/look).

File: pymc3/distributions/dist_math.py [hunk 3]
```diff
@@ -160,24 +160,37 @@ def MvNormalLogp():
...
+        # add inplace=True when/if impletemented by Theano
+        cholesky = slinalg.Cholesky(lower=True, on_error="nan")
+        cov = cholesky(cov)
+        # The Cholesky op will return NaNs if the cov is not positive definite
+        # -- checking the first value is sufficient
+        ok = ~tt.isnan(cov[0,0])
+        # will all be NaN if the Cholesky was no-go, which is fine
+        diag = tt.ExtractDiag(view=True)(cov)
+    else:
+        diag = tt.ExtractDiag(view=True)(cov)
```
Comment (ferrine) [inline pymc3/distributions/dist_math.py]: Good point to use view

File: pymc3/distributions/dist_math.py [hunk 4]
```diff
@@ -160,24 +160,37 @@ def MvNormalLogp():
...
-    cholesky = Cholesky(nofail=True, lower=True)
+    solve_lower = slinalg.Solve(A_structure='lower_triangular', overwrite_b=True)
+    solve_upper = slinalg.Solve(A_structure='upper_triangular', overwrite_b=True)
 
     n, k = delta.shape
-    n, k = f(n), f(k)
-    chol_cov = cholesky(cov)
-    diag = tt.nlinalg.diag(chol_cov)
-    ok = tt.all(diag > 0)
+    n = f(n)
```
Comment (aseyboldt) [inline pymc3/distributions/dist_math.py]: I know this wasn't your idea, but I like `floatX` a lot more. I had to go search for the definition of this.

File: pymc3/distributions/multivariate.py [hunk 5]
```diff
@@ -20,133 +21,93 @@
...
-            self.tau = tau
-            self._n = self.tau.shape[-1]
-        else:
-            self.k = chol.shape[0]
+            self.cov = tt.as_tensor_variable(cov)
+            try:
+                deltaop = MvNormalLogp()
+                def deltalogp(delta):
+                    return deltaop(self.cov, delta)
+                self.deltalogp = deltalogp
```
Comment (aseyboldt) [inline pymc3/distributions/multivariate.py]: -> `self._delta_logp`

File: pymc3/distributions/multivariate.py [hunk 6]
```diff
@@ -1235,18 +1262,20 @@ def _setup_matrices(self, colcov, colchol, coltau, rowcov, rowchol, rowtau):
             self.n = colcov.shape[0]
             self._colcov_type = 'cov'
             colcov = tt.as_tensor_variable(colcov)
-            if colcov.ndim != 2:
-                raise ValueError('colcov must be two dimensional.')
-            self.colchol_cov = cholesky(colcov)
+            try:
+                self.colchol_cov = cholesky(colcov)
```
Comment (aseyboldt) [inline pymc3/distributions/multivariate.py]: -> `self._colchol_cov`

File: pymc3/distributions/multivariate.py [hunk 7]
```diff
@@ -1206,24 +1230,27 @@ def _setup_matrices(self, colcov, colchol, coltau, rowcov, rowchol, rowtau):
...
-            raise ValueError('rowtau not supported at this time')
+            raise ValueError('`rowtau` not supported at this time')
             self.m = rowtau.shape[0]
             self._rowcov_type = 'tau'
             rowtau = tt.as_tensor_variable(rowtau)
-            if rowtau.ndim != 2:
-                raise ValueError('rowtau must be two dimensional.')
-            self.rowchol_tau = cholesky(rowtau)
+            try:
+                self.rowchol_tau = cholesky(rowtau)
```
Comment (aseyboldt) [inline pymc3/distributions/multivariate.py]: -> `self._rowchol_tau`

File: pymc3/distributions/multivariate.py [hunk 8]
```diff
@@ -20,133 +21,93 @@
...
+                    logp = norm - 0.5 * quaddist - logdet
+                    return ifelse(ok, floatX(logp), floatX(-np.inf * tt.ones_like(logp)))
+                self.deltalogp = deltalogp
+            except ValueError:
+                raise_from(ValueError('`tau` must be two dimensional.'), None)
+            self.tau = tau
 
-    def _quaddist_cov(self, delta):
-        return self._quaddist_chol(delta)
+class _QuadFormBase(Continuous, _CovSet):
```
Comment (aseyboldt) [inline pymc3/distributions/multivariate.py]: We can use `MvNormalLogp` _only_ in `MvNormal.logp_sum`, because it returns the sum of all logp values, not the elementwise logp. (we need the elemwise logp in `pm.compare` etc), so we still need `_quaddist` in MvNormal and StudentT.

File: pymc3/distributions/dist_math.py (hunk 4 above)
Comment (gBokiau) [reply pymc3/distributions/dist_math.py]: True. It's a mess though, I just seem to be wrapping FloatX around *everything*

File: pymc3/distributions/multivariate.py (hunk 8 above)
Comment (gBokiau) [reply pymc3/distributions/multivariate.py]: Ah! That's the bit I was missing! Ok, and do we *want* a GraphOp for the regular elementwise logp?

**Comment by aseyboldt** [comment]
Nice work.
A couple of benchmarks that compares performance in cases like `shape=(n)` and `shape=(1000, n)`, with chol and cov would be great.

File: pymc3/distributions/multivariate.py (hunk 8 above)
Comment (aseyboldt) [reply pymc3/distributions/multivariate.py]: I wrote that some time ago because at least back then it was quite a bit faster. But this might have changed in recent theano versions.

Comment (aseyboldt) [reply pymc3/distributions/multivariate.py]: Sorry, misread your question. The GraphOp thing is faster for the sum of logp. I don't know for sure if the same thing would work for the elemwise logp, but if I remember correctly, it doesn't. That's why MvNormalLogp isn't used at the moment. We only added `logp_sum` recently and I haven't looked into this since then.

File: pymc3/gp/gp.py [hunk 9]
```diff
@@ -104,13 +104,13 @@ def __init__(self, mean_func=Zero(), cov_func=Constant(0.0)):
...
-        chol = cholesky(stabilize(self.cov_func(X)))
+        cov = stabilize(self.cov_func(X))
         shape = infer_shape(X, kwargs.pop("shape", None))
         if reparameterize:
             v = pm.Normal(name + "_rotated_", mu=0.0, sd=1.0, shape=shape, **kwargs)
-            f = pm.Deterministic(name, mu + tt.dot(chol, v))
+            f = pm.Deterministic(name, mu + tt.dot(cholesky(cov), v))
         else:
-            f = pm.MvNormal(name, mu=mu, chol=chol, shape=shape, **kwargs)
+            f = pm.MvNormal(name, mu=mu, cov=cov, shape=shape, **kwargs)
```
Comment (bwengals) [inline pymc3/gp/gp.py:117]: Doesn't the cholesky decomp of `cov` get computed twice here?  Once on line 111 and again inside MvNormal?  Or is theano smart enough to not do that?

Comment (gBokiau) [reply pymc3/gp/gp.py:117]: There's an `if else` there so those are two separate branches and only either one of them would be computed, if I'm not mistaken?

Comment (bwengals) [reply pymc3/gp/gp.py:117]: whoops!  you're absolutely right

Comment (bwengals) [reply pymc3/gp/gp.py:117]: What's the reasoning for delegating cholesky decomp to MvNormal/StudentT?  Now sometimes the cholesky is computed within this file, and sometimes in MvNormal, potentially with different settings

Comment (gBokiau) [reply pymc3/gp/gp.py:117]: I'm trying to minimize the number of ways and places the same thing is done. In the case of line 111 there's currently no other way.
There might be some computational advantage in having the corr/Cholesky/tau matrices operations living in a separate GraphOp, in which case it would best be managed from a single point. Benchmarking that is the next thing to do/find out once *this* works, which shouldn't be long now.

Eventually, ideally, the Cholesky aspect would be abstracted away entirely in model definitions (triangular decomposition etc. should really happen in the background IMHO).

**Comment by gBokiau** [comment]
Ok, in terms of functionality this is getting there. What remains are:
- ~~Float32 and other dtype errors. I've solved what I could but am wary of sparkling more floatX's around. Reviews of the remaining errors would be very helpful.~~ Mostly "solved" by adding FloatX's in the tests. Don't know if that's ok.
- ~~Numerical errors. Some of those might be expected (I've noticed tests for the tau parameter are commented out elsewhere, so maybe that's as good as it gets?). Others are more off so the formulas need double checking.~~ Solved, was due to omissions.
- Missing algebra: the gradient for the sum of log probabilities of MvNormal(tau=tau), and, if these have a closed solution, the sum of log probabilities of MvStudent() and its gradient.

In terms of code and structure, there's room for more elegant and transparent approaches, however this should be enough to benchmark against existing code first, and more elegant solutions later.

Inference on `MvGaussianRandomWalk` would be a good place to benchmark against, as its implementation here leverages most of the tricks.

File: pymc3/gp/gp.py (hunk 9 above)
Comment (gBokiau) [reply pymc3/gp/gp.py:117]: I did notice a few remaining inconsistencies tho, now amended. 

And at second thought… the MvNormal implementation (with any of chol, cov, tau) does sanity checks that are entirely superfluous in the context of GP. As, in fact, does theano and, in turn, scipy. So there's some, albeit tiny, overhead.

So in theory there's some argument for a more distinct implementation when the chol/cov are known to be positive definite. In practice though, I'm not sure that it would be worth the effort.

Comment (aseyboldt) [reply pymc3/gp/gp.py:117]: "known to be positive definite" can be a bit tricky. It is entirely possible that a matrix is posdef, but the cholesky decomposition still fails, because there can be numerical issues if some eigenvalues are very small.

File: pymc3/distributions/dist_math.py [hunk 10]
```diff
@@ -139,19 +139,85 @@ def log_normal(x, mean, **kwargs):
...
+       Cholesky replacement if it is not, along with a boolean to assert whether
+       replacement was needed and, optionally, the log of the determinant of
+       either the real Cholesky or its replacement."""
+    is_cholesky = (mode == 'chol')
+    w_ldet = return_ldet
+    # add inplace=True when/if impletemented by Theano
+    cholesky = slinalg.Cholesky(lower=True, on_error="nan")
+
+    # Check if a given Cholesky is positive definite
+    def check_chol(cov):
```
Comment (aseyboldt) [inline pymc3/distributions/dist_math.py]: I'm not sure we actually need this check. The cholesky decomposition is only defined for positive definite matrices in the first place. So we could argue that it's the users fault if they pass in something that is not actually a cholesky matrix. (The return vals of the cholesky op should always have a positive diagonal, it couldn't take the square root otherwise)

File: pymc3/distributions/multivariate.py [hunk 11]
```diff
@@ -20,63 +20,79 @@
...
+            errot = '`{}` must be two dimensional.'.format(self._cov_type)
+            raise_from(ValueError(error), None)
+
+        self.mu = tt.as_tensor_variable(mu)
+
+    def _repr_cov_params(self):
+        name = get_variable_name(getattr(self, self._cov_type))
+        return r'\mathit{{{}}}={}'.format(self._cov_type, name)
+
+    def logp(self, value, dosum=False):
```
Comment (aseyboldt) [inline pymc3/distributions/multivariate.py]: I don't think we should add kwargs to `logp`. Just add both `logp` and `logp_sum` to `_QuadFormBase`.

File: pymc3/distributions/multivariate.py [hunk 12]
```diff
@@ -20,63 +20,79 @@
...
            'LKJCorr', 'LKJCholeskyCov', 'MatrixNormal']
 
 
-class _QuadFormBase(Continuous):
-    def __init__(self, mu=None, cov=None, chol=None, tau=None, lower=True,
-                 *args, **kwargs):
-        super(_QuadFormBase, self).__init__(*args, **kwargs)
-        if len(self.shape) > 2:
-            raise ValueError("Only 1 or 2 dimensions are allowed.")
+class _CovSet():
```
Comment (aseyboldt) [inline pymc3/distributions/multivariate.py]: Shouldn't this just be a function?

File: pymc3/distributions/multivariate.py [hunk 13]
```diff
@@ -20,63 +20,79 @@
...
+        super(_QuadFormBase, self).__init__(*args, **kwargs)
+        super(_QuadFormBase, self).__initCov__(cov, tau, chol, lower)
+
+        if len(self.shape) > 2:
+            raise ValueError("Only 1 or 2 dimensions are allowed.")
+
+        _cov = getattr(self, self._cov_type)
+
+        try:
+            _deltalogp = self._logphelper(self._cov_type)
```
Comment (aseyboldt) [inline pymc3/distributions/multivariate.py]: This is confusing. It is hard to see where `_loghelper` is coming from. The superclass should just pass `logphelper` in via `__init__`.

File: pymc3/distributions/multivariate.py [hunk 14]
```diff
@@ -20,63 +20,79 @@
...
-            self.chol_cov = tt.as_tensor_variable(chol)
-            self._n = self.chol_cov.shape[-1]
+            self.chol = tt.as_tensor_variable(chol)
+            if self.chol.ndim != 2:
+                raise ValueError('`chol` must be two dimensional.')
+        else:
+            self._cov_type = 'tau'
+            self.tau = tt.as_tensor_variable(tau)
+
+class _QuadFormBase(Continuous, _CovSet):
```
Comment (aseyboldt) [inline pymc3/distributions/multivariate.py]: Even if `_CovSet` is an object, it should be an attribute of `_QuadFormBase`, not a superclass.

File: pymc3/distributions/multivariate.py (hunk 11 above)
Comment (gBokiau) [reply pymc3/distributions/multivariate.py]: Yes I guess I could make a helper function to check for vector values that need to be cast to nd (which is the bit I'm trying not to repeat through this kwarg)

Comment (aseyboldt) [reply pymc3/distributions/multivariate.py]: I don't think it would have hurt to have that check at two places, but this works as well :-)

File: pymc3/distributions/dist_math.py (hunk 10 above)
Comment (gBokiau) [reply pymc3/distributions/dist_math.py]: Right. With GP no longer passing sampled chol's, only LKJ would, and LKJ wouldn't pass something with neg diagonal. I think that stands so I'll remove.

Comment (gBokiau) [reply pymc3/distributions/dist_math.py]: From the Travis tests it appears that FullRankGroup.L() does produce invalid choleskys. Maybe they could benefit from a hint of `stabilize()`.

Comment (gBokiau) [reply pymc3/distributions/dist_math.py]: @ferrine I think FullRankGroup is your domain, I'm afraid it's out of my scope of expertise. Is there a reason the L property might return non-positive or null diagonals? Could a check for that be implemented there? Would it be ok to add 1e-10's to the diagonal?

**Comment by gBokiau** [comment]
Thanks everyone for the feedback.

As this was turning in a bit of a monster, I've given up on using OpFromGraph from now, as it :
1. Didn't seem to provide noticeable speed improvements (I tried on a MvNormal with a 36x36 cholesky matrix, which arguably isn't the most testing situation). Ideally, this *might* shine on a sampling method that uses the gradient, with a MvNormal that is specifically cov parametrised (perhaps a GP process with NUTS, or highly dimensional FullRank approximation— if these use the gradients?). Feel free to revert to commit afaa3aa to try those things out, that commit was basically fully operational.
2. makes it impossible (due to a Theano limitation) to compile the model's gradient to a function if it has detached variables, as one would do e.g. to compile an observation-less model's gradient to a theano function in order to obtain callable that can be fed arbitrary observations (eg as the test for model.dlogp2 does).

With that aside, what this PR initially set out to do is now done. This commit trail being of very limited interest, would I split this up in distinct and clean PR's that each deal with one particular aspect so that these can be validated (or discussed) separately?

**Comment by aseyboldt** [comment]
> As this was turning in a bit of a monster

Yes, the multivariate stuff does that. It looks simple at first, but turns into a never-ending rabbit hole. :-)

> I've given up on using OpFromGraph from now

I agree. I think it is probably worth revisiting this, I remember quite significant speed ups in *some* cases. But this PR is perfectly fine without it.

A rebase or new PR (whatever you prefer) is probably a good idea, that would make things a bit more manageable.

Thanks for sticking with this.

File: pymc3/gp/gp.py [hunk 15]
```diff
@@ -200,9 +204,9 @@ def conditional(self, name, Xnew, given=None, **kwargs):
         """
         givens = self._get_given_vals(given)
         mu, cov = self._build_conditional(Xnew, *givens)
-        chol = cholesky(stabilize(cov))
+        cov = stabilize(cov)
         shape = infer_shape(Xnew, kwargs.pop("shape", None))
-        return pm.MvNormal(name, mu=mu, chol=chol, shape=shape, **kwargs)
+        return pm.MvNormal(name, mu=mu, cov=cov, shape=shape, **kwargs)
```
Comment (ferrine) [inline pymc3/gp/gp.py:209]: Does it matter @bwengals?

**Comment by ferrine** [comment]
need rebase

File: pymc3/gp/gp.py (hunk 9 above)
Comment (bwengals) [reply pymc3/gp/gp.py:117]: yup, this is the case for GPs, which is why `stabilize` is used.  The covariance functions are posdef, but computing the Cholesky often fails for numerical reasons.

File: pymc3/gp/gp.py (hunk 15 above)
Comment (bwengals) [reply pymc3/gp/gp.py:209]: I don't think it does, but why move Cholesky to MvNormal?  i feel it does make the implementation a tiny(!) bit less clear, from the standpoint of someone reading the code.  Is there a benefit?

Comment (gBokiau) [reply pymc3/gp/gp.py:209]: With the graphOp implementation -- which is *not* implemented here after all, but could/might be at some point when and if it proves satisfactory -- the benefit was that this operation happened in the 'subgraph', thus decluttering the total graph, speeding up compilation and freeing up a tiny bit of memory in the process.

From a style point of view, explicitly calling the Cholesky decomposition in the source each time (a good 8 or so times?) might lend the wrong impression that this is the preferred/optimal way to configure a Mv. In truth, Quadbase takes care of the decomposition when given a cov, so either way of configuring it is equally fine. I would find it more coherent to just pass whatever you're working with, as a best practice, since it's shorter and more intuitive.

This is entirely personal though, and has neither computational cost nor benefit. There's also the argument that `stabilize` only makes sense in the context of Cholesky decomposition, and that as it stands here, that context is lost. So I don't mind reverting to the way it was in that regard.

**Comment by gBokiau** [comment]
Indeed it does. Also, embarrassingly, I had neither tested nor considered multi-chain, and the function constructors don't pickle. So I've started another approach, which I hope to PR soon.

**Comment by gBokiau** [comment]
@ferrine I also wanted your take on wether you saw any cost to stabilizing the cov matrix in FullRankGroup, as it appears it can't always be decomposed, thus requiring otherwise unnecessary checks. i.e. IIRC at initialisation it starts off with a zero matrix that thus can't be decomposed.

**Comment by ferrine** [comment]
Hope I've fixed random Travis fails on master branch in a recent PR

**Comment by ferrine** [comment]
I saw it, good point

File: pymc3/gp/gp.py (hunk 15 above)
Comment (bwengals) [reply pymc3/gp/gp.py:209]: Ah ok, it is fine with me either way.  Just wanted to know what the reasoning was.  Thanks!

**Comment by twiecki** [comment]
@gBokiau What's the status of this? Would be great to merge it. Needs a rebase.

**Comment by gBokiau** [comment]
Apologies for the delay. I am yet to find a simple alternative to circumvent the pickling/parallelisation issue caused by the function constructors I used here, while keeping things tidy. I might be overlooking a really obvious hack though.

Also, I'm growing ever more convinced that there would be substantial value to a more thorough rewrite still of Mv's in general. My general idea is that the 'kernels' of Mv functions (LKJ etc) should compute the Mahalanobis distances themselves, instead of doing this in the `logp` of the parent distribution (as is now the case), or through intermediaries (as in this PR).

To give an example, if using `LKJCorr ` in a `MvNormal`, the procedure to add the unit diagonal and compute the Cholesky factor is done twice: once here https://github.com/pymc-devs/pymc3/blob/master/pymc3/distributions/multivariate.py#L1091, and once respectively in passing it as a parameter and subsequently in `MvNormal.logp`. Considering the cost of decomposing the matrix, this is hardly ideal.

To illustrate, the goal would be to be able to do this : 
```
kernel = LKJCorr('kernel', eta=1)
observations = MvNormal('name', kernel=kernel, observed=data)
```
Ideally the dimensions of the kernel would be inferred automatically.

Perhaps even this:
```
spherical_kernel = SphericalKernel(a=2)
observations = MvNormal('name', kernel=spherical_kernel, observed=data)
```
would be more 'natural' than declaring regular `Normal(sd=2)` when you essentially mean an uncorrelated bivariate normal (but this is debatable).

More to the point:
```
precision_kernel = PrecsionKernel(np.array([[1., 2.],[2.,1.]]))
observations = MvNormal('name', kernel=precision_kernel, observed=data)
```
…would be a new way to declare `tau`, and would be what happens behind the scene if that is how it is declared.

In short, in terms of architecture, I think `quaddist` should move from the distributions to the kernels. They would be a family of classes of their own, which can but are not required to be RVs.

Yet the implications are rather considerable. I'm struggling to find a promising a starting point to get a proof of concept going.

**Comment by twiecki** [comment]
Did you mean to close this @gBokiau ?

**Comment by gBokiau** [comment]
Oops, that was accidental — but perhaps appropriate?

**Comment by junpenglao** [comment]
I think it is a bit of a waste of all the hard work. However, I am not convinced that the distance should be computed in the kernel instead of the parent Mv Distribution. I agree that decomposition should be done only once, and a bit surprised that it is currently computed twice (not the case if you pass `chol` to MvNormal right?)

**Comment by gBokiau** [comment]
> not the case if you pass chol to MvNormal right?

Correct, only with `cov=packed(LKJCorr)+np.eye(k)`.  This is because LKJCorr pseudo-computes the cholesky to check if the sampled values are P.D., when computing logp. The result should instead obviously be saved and reused by MvN/MvT.

**Comment by junpenglao** [comment]
I see... hmmm this is a good point you raise, so what you are doing now is trying to combine the P.D. checking with logp computation by doing the decomposition midway? And you think it is better to do it when the cov is being declared? 

And I imagine that's what `SphericalKernel` etc are for? A class that make sure P.D. and also do the necessary decomposition. But what about if you are combining different kernels such as in `gp`?

**Comment by gBokiau** [comment]
In that particular case, I think I would move towards having LKJCorr always save results Cholesky decomposed with unit diagonal (ie transformed), since they're being computed anyway and are not as costly to reverse-transform. However it's not quite clear to me when/how one would use LKJCorr without a unit diagonal, and if/how to accommodate for that use case.

The more general case for having the kernels compute the distance is basically twofold. 
First, we need three approaches:
- precision-based: distance obtained by simple dot product
- cholesky factor-based: distance obtained by solving triangular
- cov/corr-based: need to first check if P.D., then cholesky decompose, than as above.

Inheritance would be a more elegant approach than intricate switch loops, as both here and in master.

Second, a cov/corr kernel is decomposed as many times as it is used. While definitely a fringe case, one might model a mixture between multiple MvN with a shared covariance matrix but different means. Certainly in the case of GP's it seems plausible that kernels would be reused in different RV's.

Thankfully, I do think 'fixed' correlation matrices (ie np.array's) are only decomposed during model computation, yet that would be worth double checking as well.

**Comment by gBokiau** [comment]
@junpenglao not sure by what you mean by midway — probably because I haven't yet figured it all out myself.

Another way I look at it is that the GP kernels would : 
1) be computed 'later' (not in model declaration), presumably by the distribution, trough `draw_values()` calls
2) once drawn/sampled, return 'augmented' variables, so as to have `transform`,  `k`, and `quaddist` methods, the two former so that the distribution can generate random samples, the latter so that it can compute logp. Plus the properties they already have, ie diagonal and full.
3) under the hood, either be cholesky or tau matrices, or implement/inherit methods to deliver those, which would be performed only once.

I'm the first to admit the implications are intricate. Additions and multiplications could be handled as they are currently for GP kernels. Some (most?) of these ops could be implemented directly [on the Cholesky factors](https://en.wikipedia.org/wiki/Cholesky_decomposition#Updating_the_decomposition).

There's also the matter of wether it would be worth implementing this on Whishart etc., or rather to drop those.

**Comment by twiecki** [comment]
> There's also the matter of wether it would be worth implementing this on Whishart etc., or rather to drop those.

We can probably drop those, I don't think anyone uses those anymore (and they shouldn't).

**Comment by gBokiau** [comment]
In the same vein, my impression is that `LKJCorr()` offers no advantage over `LKJCholeskyCov(sd=1)`. Discarding that as well would considerably alleviate the specs and solve the problem above.

I seem to recall mentions of computing GP kernels directly in "Cholesky space". Has anyone come across that? It's a tough one to find the right keywords to search for.

**Comment by aseyboldt** [comment]
I don't think the current parametrization of `LKJCorr` is great, but it can't be replaced that easily with `LKJCholeskyCov`, because that only works if the variances are free random variables. So `LKJCholeskyCov(sd=1)` doesn't actually work. This is because we have one free variable per non-zero entry in the chol matrix, but no free variable for the variance (which is just the norm of one row of the chol matrix). If you want a LKJ Correlation (rather than covariance) matrix, you could right now just use an arbitrary dist for the sd, and then rescale the covariance matrix to a correlation matrix and use that. Using a stick breaking approach we could also parameterize the correlation matrix directly, but that isn't implemented. 
Or maybe skip the whole cholesky decomposition and go for an eigenvalue decomposition instead. There, we could use gibs rotations or householder transforms of skew symmetric matrices to parameterize the orthogonal matrix.

**Comment by junpenglao** [comment]
@gBokiau How do you feel about this PR? I think it is still worth to pursuit it - maybe not over complicate it and clean it up so that the current test pass?

**Comment by twiecki** [comment]
@gBokiau I'm going to close this as I think your other approach with making smaller PRs is the way to go here (let me know if you would like to continue here, however, and we can certainly reopen it).

**Comment by gBokiau** [comment]
Thanks @twiecki, sounds good. I intend to see wether the logp approach fares better under @aseyboldt 's new parallelism approach shortly.
